proposals/Exceptions.md

```diff
@@ -38,125 +39,136 @@ and the instructions *throw* and *rethrow*.
 
-**TODO** Define the order that the data fields of the exception should appear on
-the value stack.
+### Try and catch blocks.
 
-When an exception is thrown, the corresponding values are popped off the value
-stack, and stored internally for access when the exception is caught.
+A _try_ block defines a list of instruction that may need to catch exceptions
+and/or clean up state when an exception is thrown.  Like other higher-level
+constructs, a try block begins with a `try` instruction, and ends with an `end`
+instruction. That is, a try block is sequence of instructions having the
+following form:
+
+```
+try
+  instruction*
+catch i
+  instruction*
+catch j
+  instruction*
+...
+catch n
+  instruction*
+else
+    instruction*
+end
+```
+
+A try block may contain one or more catch blocks, and all but the last catch
+block must begin with a`catch` instruction. The last catch block can begin with
+either a `catch` or `else` instruction. The `catch`/`else` instructions (within
+the try construct) are called the _catching_ instructions.
+
+The _body_ of the try block is the list of instructions before the first
+catching instruction. The _body_ of each catch block is the sequence of
+instructions following the corresponding catching instruction, and the next
+catching instruction (or the `end` instruction if it is the last catching
+block).
+
+The `catch` instruction has an exception tag associated with it. The tag
+identifies what exceptions it can catch. That is, any exception created with the
+corresponding exception tag. Catch blocks that begin with a `catch` instruction
+are considered a _tagged_ catch block.
+
+The last catch block of an exception can be a tagged catch block. Alternatively,
+it can begin with the `else` instruction. If it begins with the `else`
+instruction, it defines the _default_ catch block. The default catch block has
+no exception type, and is used to catch all exceptions not caught by any of the
+tagged catch blocks.
+
+Try blocks, like a control-flow blocks, have a _block type_. The block type of a
+try block defines the values yielded by the evaluation the try block when either
+no exception is thrown, or the exception is successfully caught by one of its
+catch blocks, and the instructions within the catch block can recover from the
+throw.
+
+In the initial implementation, try blocks may only yield 0 or 1 values.
 
-If the throw appears within a try block, the value stack is popped back to
-the size the value stack had when the try block was initially entered.
+### Throws
 
-If a throw occurs within a function body, and it doesn't have an enclosing try
-block, the throw continues up the call stack until either an enclosing try block
-is found, or the call stack is flushed. If the call stack is flushed, execution
-is terminated.
+The _throw_ instruction has a single immediate argument, an exception tag.  The
+exception tag is used to define the constructed exception. The arguments for the
```

Comment:
Reworded this sentence to "When an exception is thrown, the exception is allocated and the values on the stack (corresponding to the type signature) are popped off and assigned to the allocated exception.